state/protocol/badger/snapshot.go

```diff
@@ -265,24 +265,47 @@ func (s *Snapshot) SealedResult() (*flow.ExecutionResult, *flow.Seal, error) {
 	return result, seal, nil
 }
 
-func (s *Snapshot) SealingSegment() ([]*flow.Block, error) {
+// SealingSegment will walk through the chain backward until we reach the block referenced
+// by the latest seal and build a SealingSegment. As we visit each block we check each execution
+// receipt in the blocks payload to make sure we have a corresponding execution result, any execution
+// results missing from blocks are stored in the SealingSegment.ExecutionResults field.
+func (s *Snapshot) SealingSegment() (*flow.SealingSegment, error) {
 	seal, err := s.state.seals.ByBlockID(s.blockID)
 	if err != nil {
-		return nil, fmt.Errorf("could not get seal for sealing segment: %w", err)
+		return nil, fmt.Errorf("could not get seal for sealing seg: %w", err)
 	}
 
 	// walk through the chain backward until we reach the block referenced by
 	// the latest seal - the returned segment includes this block
-	var segment []*flow.Block
+	segment := flow.NewSealingSegment()
+	includedResults := make(map[flow.Identifier]*flow.ExecutionResult)
```

Comment:
It's better to create a `SegmentBuilder` that holds this lookup internally, and in the end returns the segment.

The `SegmentBuilder` will take a method to lookup missing result by result ID, and when passing a block, it will go through the each receipt, and check if its result are known, if not, it will look it up. 
```
type SegmentBuilder struct {
  resultFinder      func(resultID flow.Identifier) (*flow.ExecutionResult, error)
  includedResults   map[flow.Identifier]*flow.ExecutionResult
  blocks  []*flow.Block
  results []*flow.ExecutionResult
}

// find the missing result to be added to 
func (b *SegmentBuilder) AddBlock(b *flow.Block) error {
}
func (b *SegmentBuilder) BuildSegment() (*Segment, error) {
}

```

And the BuildSegment function should validate that the highest block should include a seal that seals the lowest block, otherwise an error should be returned to indicate the Segment is incomplete.

Reply:
This is more of a style preference, is there a builder API I can follow that's consistent throughout the repo ?

Reply:
You can refer to the golang's [string builder](https://pkg.go.dev/strings#Builder), is similar to the design the suggested here. where you provide a block, and it takes care of looking up results. 

The nice thing with this is that the caller (protocol state) doesn't need to know how to build a segment. And it's also easier to unit test the building logic in SegmentBuilder

Reply:
I was hoping for something more flow-go specific, I can add a generic builder that is fine. Just to be clear essentially we will be moving the logic from [Snapshot.SealingSegment()](https://github.com/onflow/flow-go/blob/master/state/protocol/badger/snapshot.go#L268) and replacing it with this new builder pattern?